pkg/common/cns-lib/volume/listview.go

```diff
@@ -0,0 +1,360 @@
+	if err != nil {
+		return logger.LogNewErrorf(log, "failed to create a govmomi client for listView. error: %+v", err)
+	}
+	client.Timeout = noTimeout
+	t.govmomiClient = client
+	return nil
+}
+
+// listenToTaskUpdates is a long-running goroutine
+// that uses a property collector to listen for task updates
+// CSI ops add CNS tasks to listview and wait for a response from CNS
+// when update(s) are received by the property collector,
+// it spawns a new goroutine to process each task update and return the result to the caller
+func (t *ListViewImpl) listenToTaskUpdates() {
+	log := logger.GetLogger(t.ctx)
+	filter := getListViewWaitFilter(t.listView)
+	// we need to recreate the listView and the wait filter after any error from vc
+	// for the first iteration we already have the listView and filter initialized
+	recreateView := false
+	for {
+		// calling Connect at the beginning to ensure the current session is neither nil nor NotAuthenticated
+		if err := t.isClientValid(); err != nil {
+			log.Errorf("failed to connect to vCenter. err: %v", err)
+			time.Sleep(waitForUpdatesRetry)
+			continue
+		} else {
+			log.Infof("connection to vc successful")
+		}
+
+		if recreateView {
+			log.Info("re-creating the listView object")
+			err := t.createListView(t.ctx, nil)
+			if err != nil {
+				log.Errorf("failed to create a ListView. error: %+v", err)
+				continue
+			}
+
+			filter = getListViewWaitFilter(t.listView)
+			recreateView = false
+		}
+
+		log.Info("Starting listening for task updates...")
+		pc := property.DefaultCollector(t.govmomiClient.Client)
+		err := property.WaitForUpdates(t.ctx, pc, filter, func(updates []types.ObjectUpdate) bool {
+			log.Debugf("Got %d property collector update(s)", len(updates))
+			for _, update := range updates {
+				for _, prop := range update.ChangeSet {
+					log.Debugf("Got update for object %v properties %v", update.Obj, prop)
+					// we don't need a lock at this line as we aren't accessing any map item
+					go t.processTaskUpdate(prop)
+				}
+			}
+
+			// this return value is used by the WaitForUpdates method.
+			// we only want this true while running the unit tests so the test can finish
+			return t.shouldStopListening
+		})
+		// if property collector returns any errors,
+		// we want to immediately return a fault for all the pending tasks in the map
+		// note: this is not a task error but an error from the vc
+		if err != nil {
+			log.Errorf("WaitForUpdates returned err: %v for vc: %+v", err, t.virtualCenter)
+			recreateView = true
+
+			// in case of failure, we return failure for the tasks present in the map
+			// reading from a map requires acquiring a reader lock
+			// if a writer lock is called first, reader lock will block until the writer lock is released & vice-versa
+
+			t.counter.RLock()
+			for taskMoRef := range t.counter.taskMap {
+				prop := types.PropertyChange{
+					Val: types.TaskInfo{
+						Task:  taskMoRef,
+						State: types.TaskInfoStateError,
```

Comment:
This sort of implies that there is an error in the task itself, which is not the case, can we have any other way to indicate property collector/infra failure rather than specific task failure.

Reply:
If you look at `processTaskUpdate(),` the error from the task is returned as part of the `Result` struct. 

So, at the place where we wait, the error we got from the ListView property collector will be assigned to the `err` at the below line. 
```
		var taskInfo *vim25types.TaskInfo
		if m.tasksListViewEnabled {
			taskInfo, err = m.waitOnTask(ctx, task.Reference())
		} else {
			taskInfo, err = cns.GetTaskInfo(ctx, task)
		}
```

This is the same with the current `cns.GetTaskInfo` as can be seen from https://github.com/vmware/govmomi/blob/master/task/wait.go#L129-L140